exercises/book/A3-traits-generics/mod.md

```diff
@@ -0,0 +1,151 @@
+
+First off, we will implement the [`IntoIterator`](https://doc.rust-lang.org/std/iter/trait.IntoIterator.html) and [`Iterator`](https://doc.rust-lang.org/std/iter/trait.Iterator.html) traits. Go ahead and uncomment the `it_iters` test case. Let's define a new type:
+
+
+```rust
+pub struct LocalStorageVecIter<T, const N: usize> {
+    vec: LocalStorageVec<T, N>,
+    counter: usize,
+}
+```
+
+This is the type we'll implement the `Iterator` trait on. You'll need to specify the item this `Iterator` implementation yields, as well as an implementation for `Iterator::next`, which yields the next item. You'll be able to make this easier by bounding `T` to `Default` when implementing the `Iterator` trait, as then you can use the [`std::mem::take`](https://doc.rust-lang.org/std/mem/fn.take.html) function to take an item from the `LocalStorageVec` and replace it with the default value for `T`.
+
+Take a look at the list of methods under the ['provided methods' section](https://doc.rust-lang.org/std/iter/trait.Iterator.html). In there, lots of useful methods that come free with the implementation of the `Iterator` trait are defined, and implemented in terms of the `next` method. Knowing in the back of your head what methods there are, greatly helps in improving your efficiency in programming with Rust. Which of the provided methods can you override in order to make the implementation of `LocalStorageVecIter` more efficient, given that we can access the fields and methods of `LocalStorageVec`?
+
+Now to instantiate a `LocalStorageVecIter`, implement the [`IntoIter`] trait for it, in such a way that calling `into_iter` yields a `LocalStorageVecIter`.
+
+### A3.2.E `AsRef` and `AsMut`
+[`AsRef`](https://doc.rust-lang.org/std/convert/trait.AsRef.html) and [`AsMut`](https://doc.rust-lang.org/std/convert/trait.AsMut.html) are used to implement *cheap* reference-to-reference coercion. For instance, our `LocalStorageVec<T, N>` is somewhat similar to a slice `&[T]`, as both represent a contiguous series of `T` values. This is true whether the `LocalStorageVec` buffer resides on the stack or on the heap. 
+
+Uncomment the `it_as_refs` test case and implement `AsRef<[T]>` and `AsMut<[T]>`.
+
+### A3.2.F `Index`
+To allow users of the `LocalStorageVec` to read items or slices from its buffer, we can implement the [`Index`](https://doc.rust-lang.org/std/ops/trait.Index.html) trait. This trait is generic over the type of the item used for indexing. In order to make our `LocalStorageVec` versatile, we should implement: 
+
+- `Index<uize>`, allowing us to get a single item by calling `vec[1]`;
+- `Index<RangeTo<usize>>`, allowing us to get the first `n` items (excluding item `n`) by calling `vec[..n]`;
+- `Index<RangeFrom<usize>>`, allowing us to get the last `n` items by calling `vec[n..]`;
+- `Index<Range<usize>>`, allowing us to get the items between `n` and `m` items (excluding item `m`) by calling `vec[n..m]`;
+
```

Comment:
Is it really necesary to implement it four times? If done properly this will end up with a lot of boilerplate (unless they immediately do part H, but that is pretty advanced stuff.

Reply:
The idea is that students get an felling of how blanket implementations can be used to reduce boilerplate.